File: ProgressBarThreadTuto/app/src/main/java/com/example/progressbarthreadtuto/MainActivity.java
package com.example.progressbarthreadtuto;

import android.os.Bundle;
import android.widget.ProgressBar;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    private static final int PROGRESS=0x1;//뭔지 모르겠다.
    private ProgressBar mProgress;//Progress Bar
    private int mProgressStatus = 0;//Progress Bar 진행도
    int i = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mProgress = findViewById(R.id.progressBar);
        final Thread t = new Thread(new Runnable() {//스레드에 Runable 구현
                    @Override
                    public void run() {//run 메소드 오버라이딩
                        while (mProgressStatus<100){//진행도 100될 때 까지 반복
                            try {
                                Thread.sleep(1000);//1초 동안 스레드 슬립
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                            mProgressStatus = i++;//깨면 진행도 1 상승
                            mProgress.post(new Runnable() {//UI 스레드에 Post할 Runnable 구현
                                @Override
                                public void run() {
                                    mProgress.setProgress(mProgressStatus);//진행도 set
                                }
                            });
                        }
                    }
                });
                t.start();//thread실행
    }
}

File: ServiceTest/app/src/main/java/com/example/servicetest/MainActivity.kt
package com.example.servicetest

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.RemoteException
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    private var tvCounter: TextView? = null
    private var btnStart: Button? = null
    private var btnStop: Button? = null
    private var binder: IMyCounterService? = null

    /**
     * Activity가 Service를 호출합니다.
     * 이 때, 꼭 binder가 필요합니다.
     */
    private val connection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            /**
             * Service가 가지고있는 binder를 전달받는다.
             * 즉, Service에서 구체화한 getCount() 메소드를 받았습니다.
             */
            binder = IMyCounterService.Stub.asInterface(service)
        }

        override fun onServiceDisconnected(name: ComponentName) {}
    }
    private var intent: Intent? = null
    private var running = true
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        tvCounter = findViewById<View>(R.id.tvCounter) as TextView
        btnStart = findViewById<View>(R.id.btnStart) as Button
        btnStart!!.setOnClickListener {
            intent = Intent(this@MainActivity, MyCounterService::class.java)
            bindService(intent, connection, BIND_AUTO_CREATE)
            running = true
            Thread(GetCountThread()).start()
        }
        btnStop = findViewById<View>(R.id.btnStop) as Button
        btnStop!!.setOnClickListener {
            unbindService(connection)
            running = false
        }
    }

    /**
     *
     */
    private inner class GetCountThread : Runnable {
        // binder에서 count가져와서 set시키려면 handler 필요
        private val handler = Handler()
        override fun run() {
            while (running) {
                if (binder == null) {
                    continue
                }
                handler.post {
                    try {
                        tvCounter!!.text = binder!!.count.toString() + ""
                    } catch (e: RemoteException) {
                        e.printStackTrace()
                    }
                }

                // 0.5초 텀을 준다.
                try {
                    Thread.sleep(500)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }
    }
}